### PR TITLE
Initial Flink DataSet API support

### PR DESCRIPTION
**Pull request by jedrz**
Assumptions:
* make as little as possible code changes in existing streaming api support
* should be able to upload a jar via nk to flink cluster and execute it using DataSet api
* metrics are not yet implemented
* joins are not yet implemented
* custom nodes support is missing

TODO in next changes:
* extract common code
* implement omitted features
* add more (integration) tests
* support native DataSet operations (via custom nodes?)

Initial support for #272

### CODE REVIEW

File: engine/flink/management/src/main/scala/pl/touk/nussknacker/engine/management/FlinkRestManager.scala [hunk 1]
```diff
@@ -23,8 +23,9 @@ import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent.Future
 
 
-class FlinkRestManager(config: FlinkConfig, modelData: ModelData)(implicit backend: SttpBackend[Future, Nothing, NothingT])
-    extends FlinkProcessManager(modelData, config.shouldVerifyBeforeDeploy.getOrElse(true)) with LazyLogging {
+class FlinkRestManager(config: FlinkConfig, modelData: ModelData, mainClassName: String = "pl.touk.nussknacker.engine.process.runner.FlinkProcessMain")
```
Comment (mproch) [inline engine/flink/management/src/main/scala/pl/touk/nussknacker/engine/management/FlinkRestManager.scala]: Maybe change name here (and in other appropriate places) to FlinkStreamingRestManager etc?

File: engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/runner/BatchFlinkProcessMain.scala [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        logger.error("Unhandled error", ex)
+        throw ex
+    }
+  }
+
+  private def prepareRegistrar(processConfigCreator: ProcessConfigCreator, config: Config): BatchFlinkProcessRegistrar = {
+    new BatchStandardFlinkProcessCompiler(processConfigCreator, config).createBatchFlinkProcessRegistrar()
+  }
+
+  private def setBuildInfo(buildInfo: String, processVersion: ProcessVersion, env: ExecutionEnvironment): Unit = {
```
Comment (mproch) [inline engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/runner/BatchFlinkProcessMain.scala]: maybe we can extract it to some mehod/class?

File: engine/util/src/main/scala/pl/touk/nussknacker/engine/util/config/ConfigEnrichments.scala [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package pl.touk.nussknacker.engine.util.config
+
+import com.typesafe.config.{Config, ConfigFactory}
+import net.ceedubs.ficus.SimpleFicusConfig
+import net.ceedubs.ficus.readers.ValueReader
+
+object ConfigEnrichments {
+
+  implicit class RichConfig(config: Config) {
```
Comment (mproch) [inline engine/util/src/main/scala/pl/touk/nussknacker/engine/util/config/ConfigEnrichments.scala]: :+1:

File: engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/BatchFlinkProcessRegistrar.scala [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+import pl.touk.nussknacker.engine.splittedgraph.splittednode.SplittedNode
+import pl.touk.nussknacker.engine.util.metrics.RateMeter
+import pl.touk.nussknacker.engine.util.{SynchronousExecutionContext, ThreadUtils}
+
+import scala.concurrent.{Await, ExecutionContext}
+import scala.language.implicitConversions
+import scala.reflect.ClassTag
+import scala.util.control.NonFatal
+
+class BatchFlinkProcessRegistrar(compileProcess: (EspProcess, ProcessVersion) => ClassLoader => CompiledProcessWithDeps,
```
Comment (mproch) [inline engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/BatchFlinkProcessRegistrar.scala]: Can we at least write somewhere what works and what does not? 
Eg. we have to think about custom transformer API and also about optimizations for SyncInterpretationFunction - I hope we can use mapPartition here

File: engine/flink/management/sample/src/main/scala/pl/touk/nussknacker/engine/management/sample/TestProcessConfigCreator.scala [hunk 5]
```diff
@@ -62,7 +68,8 @@ class TestProcessConfigCreator extends ProcessConfigCreator {
           override def serializeKey(element: Any) = null
 
           override def getTargetTopic(element: Any) = null
-        }), "Category1", "Category2")
+        }), "Category1", "Category2"),
+      "batch-file-sink" -> WithCategories(BatchFileSinkFactory, "Category1", "Category2")
```
[the remark this answers is not in the document]
Comment (jedrz) [reply engine/flink/management/sample/src/main/scala/pl/touk/nussknacker/engine/management/sample/TestProcessConfigCreator.scala]: Done. I created a separate module with test batch config.

File: engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/runner/BatchFlinkProcessMain.scala (hunk 2 above)
Comment (jedrz) [reply engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/runner/BatchFlinkProcessMain.scala]: Done.

File: engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/BatchFlinkProcessRegistrar.scala (hunk 4 above)
Comment (jedrz) [reply engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/BatchFlinkProcessRegistrar.scala]: I created one big issue #272 and linked this PR there. Also added some TODOs in appropriate places in code.

File: engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/FlinkBatchProcessRegistrar.scala [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+  }
+
+  private def usingRightClassloader(env: ExecutionEnvironment)(action: => Unit): Unit = {
+    if (!env.getJavaEnv.isInstanceOf[RemoteEnvironment]) {
+      val flinkLoaderSimulation =  FlinkUserCodeClassLoaders.childFirst(Array.empty, Thread.currentThread().getContextClassLoader, Array.empty)
+      ThreadUtils.withThisAsContextClassLoader[Unit](flinkLoaderSimulation)(action)
+    } else {
+      action
+    }
+  }
```
Comment (mproch) [inline engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/FlinkBatchProcessRegistrar.scala]: can we make it common with streaming?

File: build.sbt [hunk 7]
```diff
@@ -303,6 +304,20 @@ lazy val managementJavaSample = (project in engine("flink/management/java_sample
...
+    assemblyJarName in assembly := "managementBatchSample.jar",
+    assemblyOption in assembly := (assemblyOption in assembly).value.copy(includeScala = false, level = Level.Debug),
+    test in assembly := {},
+    libraryDependencies ++= {
+      Seq(
+        "org.apache.flink" %% "flink-scala" % flinkV % "provided",
+      )
+    }
+
+  ).dependsOn(flinkUtil, kafka, kafkaFlinkUtil, process % "runtime,test", flinkTestUtil % "test", kafkaTestUtil % "test", securityApi)
```
[the remark this answers is not in the document]
Comment (jedrz) [reply build.sbt]: Done. More unused dependencies were removed.

File: engine/flink/api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/batch/FlinkInputFormat.scala [hunk 8]
```diff
@@ -0,0 +1,30 @@
+package pl.touk.nussknacker.engine.flink.api.process.batch
+
+import org.apache.flink.api.common.io.InputFormat
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import pl.touk.nussknacker.engine.api.MethodToInvoke
+import pl.touk.nussknacker.engine.api.process.{Source, SourceFactory}
+
+import scala.reflect.ClassTag
+
+trait FlinkInputFormat[T] extends Source[T] {
```
[the remark this answers is not in the document]
Comment (jedrz) [reply engine/flink/api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/batch/FlinkInputFormat.scala]: Done.

File: engine/flink/management/src/it/scala/pl/touk/nussknacker/engine/management/FlinkBatchProcessManagerSpec.scala [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+  test("deploy process in running flink") {
+    val processName = ProcessName("batchProcess")
+
+    val version = ProcessVersion(versionId = 15, processName = processName, user = "user1", modelVersion = Some(13))
+    val process = prepareProcess(processName)
+
+    deployProcessAndWaitUntilFinished(process, version)
+
+    processVersion(processName) shouldBe Some(version)
```
[the remark this answers is not in the document]
Comment (jedrz) [reply engine/flink/management/src/it/scala/pl/touk/nussknacker/engine/management/FlinkBatchProcessManagerSpec.scala]: Done.

File: engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/FlinkBatchProcessRegistrar.scala (hunk 6 above)
Comment (jedrz) [reply engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/FlinkBatchProcessRegistrar.scala]: You're right. I should have extracted it.

File: engine/flink/process/src/test/scala/pl/touk/nussknacker/engine/process/BatchProcessTestHelpers.scala [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+    @MethodToInvoke
+    def invoke()(implicit ec: ExecutionContext): Future[String] = {
+      measuring(Future.successful {
+        internalVar
+      })
+    }
+  }
+
+  case object MonitorEmptySink extends FlinkSink {
```
[the remark this answers is not in the document]
Comment (jedrz) [reply engine/flink/process/src/test/scala/pl/touk/nussknacker/engine/process/BatchProcessTestHelpers.scala]: Done.

File: engine/flink/process/src/test/scala/pl/touk/nussknacker/engine/process/BatchProcessTestHelpers.scala [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+import pl.touk.nussknacker.engine.flink.util.service.TimeMeasuringService
+import pl.touk.nussknacker.engine.flink.util.source.FlinkCollectionInputFormat
+import pl.touk.nussknacker.engine.graph.EspProcess
+import pl.touk.nussknacker.engine.process.CommonTestHelpers.RecordingExceptionHandler
+import pl.touk.nussknacker.engine.process.compiler.FlinkBatchProcessCompiler
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.reflect.ClassTag
+
+object BatchProcessTestHelpers {
```
[the remark this answers is not in the document]
Comment (jedrz) [reply engine/flink/process/src/test/scala/pl/touk/nussknacker/engine/process/BatchProcessTestHelpers.scala]: Done.

File: engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/compiler/StubbedFlinkProcessCompiler.scala [hunk 12]
```diff
@@ -16,7 +16,7 @@ import pl.touk.nussknacker.engine.graph.node.Source
 import shapeless.syntax.typeable._
 
 abstract class StubbedFlinkProcessCompiler(process: EspProcess, creator: ProcessConfigCreator, config: Config)
-  extends FlinkProcessCompiler(creator, config, diskStateBackendSupport = false) {
+  extends FlinkStreamingProcessCompiler(creator, config, diskStateBackendSupport = false) {
```
[the remark this answers is not in the document]
Comment (jedrz) [reply engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/compiler/StubbedFlinkProcessCompiler.scala]: It's left to be implemented. There's a note in the main issue to implement tests from file for batch. Though I think it should be quiet easily doable so if you want I can try to do it in this PR :)

File: engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/compiler/FlinkProcessCompiler.scala [hunk 13]
```diff
@@ -17,7 +17,7 @@ import pl.touk.nussknacker.engine.flink.api.signal.FlinkProcessSignalSender
 import pl.touk.nussknacker.engine.flink.util.async.DefaultAsyncExecutionConfigPreparer
 import pl.touk.nussknacker.engine.flink.util.listener.NodeCountMetricListener
 import pl.touk.nussknacker.engine.graph.EspProcess
-import pl.touk.nussknacker.engine.process.FlinkProcessRegistrar
+import pl.touk.nussknacker.engine.process.FlinkStreamingProcessRegistrar
```
[the remark this answers is not in the document]
Comment (jedrz) [reply engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/compiler/FlinkProcessCompiler.scala]: Done.

File: engine/flink/management/batch_sample/src/main/scala/pl/touk/nussknacker/engine/management/batch/sample/BatchTestProcessConfigCreator.scala [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+import pl.touk.nussknacker.engine.api._
+import pl.touk.nussknacker.engine.api.exception.ExceptionHandlerFactory
+import pl.touk.nussknacker.engine.api.process._
+import pl.touk.nussknacker.engine.flink.api.process.batch.{FlinkInputFormat, FlinkInputFormatFactory, FlinkOutputFormat}
+import pl.touk.nussknacker.engine.flink.util.exception.VerboselyLoggingExceptionHandler
+import pl.touk.nussknacker.engine.util.LoggingListener
+
+import scala.reflect.ClassTag
+
+class BatchTestProcessConfigCreator extends ProcessConfigCreator {
```
[the remark this answers is not in the document]
Comment (jedrz) [reply engine/flink/management/batch_sample/src/main/scala/pl/touk/nussknacker/engine/management/batch/sample/BatchTestProcessConfigCreator.scala]: I left ProcessConfigCreator here to not introduce additional dependency on the interpreter.

[the remark this answers is not in the document]
Comment (mproch) [reply engine/flink/management/batch_sample/src/main/scala/pl/touk/nussknacker/engine/management/batch/sample/BatchTestProcessConfigCreator.scala]: aaa... ok... or we can move EmptyProcessConfigCreator to API later ;)

**Pull request merged**